File: TourismAgency/AgencyNetworking/src/main/java/protobuffprotocol/AgencyProtoWorker.java
package protobuffprotocol;

import domain.Reservation;
import domain.TravelAgent;
import domain.Trip;
import domain.validators.ValidationException;
import repository.RepoException;
import services.IAgencyObserver;
import services.IAgencyService;
import services.ServiceException;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.Socket;
import java.time.LocalTime;

public class AgencyProtoWorker implements Runnable, IAgencyObserver {

    private IAgencyService server;
    private Socket connection;

    private InputStream input;
    private OutputStream output;
    private volatile boolean connected;

    public AgencyProtoWorker(IAgencyService server, Socket connection) {
        this.server = server;
        this.connection = connection;
        try{
            output=connection.getOutputStream() ;//new ObjectOutputStream(connection.getOutputStream());
            input=connection.getInputStream(); //new ObjectInputStream(connection.getInputStream());
            connected=true;
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    @Override
    public void run() {
        while(connected){
            try {
                // Object request=input.readObject();
                System.out.println("Waiting requests ...");
                AgencyProtobufs.AgencyRequest request=AgencyProtobufs.AgencyRequest.parseDelimitedFrom(input);
                System.out.println("Request received: "+request);
                AgencyProtobufs.AgencyResponse response=handleRequest(request);
                if (response!=null){
                    sendResponse(response);
                }
            } catch (IOException e) {
                e.printStackTrace();
            }
            try {
                Thread.sleep(1000);
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        }
        try {
            input.close();
            output.close();
            connection.close();
        } catch (IOException e) {
            System.out.println("Error "+e);
        }
    }
    @Override
    public void reservationAdded(Reservation newReservation) throws ServiceException {

        System.out.println("Update new reservation...");
        try {
            sendResponse(ProtoUtils.createNewReservationResponse(newReservation));
        } catch (IOException e) {
            throw new ServiceException("Sending  new reservation response error... "+e);
        }
    }

    private AgencyProtobufs.AgencyResponse handleRequest(AgencyProtobufs.AgencyRequest request) {
        AgencyProtobufs.AgencyResponse response = null;
        switch (request.getType()){
            case LOGIN -> {
                System.out.println("LOGIN REQUEST...");
                TravelAgent user = ProtoUtils.getUser(request);
                try{
                    server.login(user,this);
                    return ProtoUtils.createOkResponse();
                } catch (ServiceException e) {
                    //connected=false;
                    return ProtoUtils.createErrorResponse(e.getMessage());
                }
            }
            case LOGOUT -> {
                System.out.println("LOGOUT REQUEST...");
                TravelAgent user = ProtoUtils.getUser(request);
                try{
                    server.logout(user,this);
                    connected=false;
                    return ProtoUtils.createOkResponse();
                } catch (ServiceException e) {

                    return ProtoUtils.createErrorResponse(e.getMessage());
                }
            }
            case ADD_USER -> {
                System.out.println("Add user request..");
                TravelAgent user = ProtoUtils.getUser(request);
                try{
                    server.addUser(user);
                    return ProtoUtils.createOkResponse();
                } catch (ServiceException | ValidationException | RepoException e) {
                    return ProtoUtils.createErrorResponse(e.getMessage());
                }
            }
            case GET_ALL_TRIPS -> {
                System.out.println("Get all trips request..");
                try {
                    Iterable<Trip> trips = server.getAllTrips();
                    return ProtoUtils.createGetAllTripsResponse(trips);
                } catch (ServiceException e) {
                    return ProtoUtils.createErrorResponse(e.getMessage());
                }
            }
            case TRIPS_BY_NAME_HOUR -> {
                System.out.println("Get filtered trips request..");
                try {
                    String s = request.getFilterTrips().getPlace();
                    LocalTime min = LocalTime.of(request.getFilterTrips().getMinHour(),0);
                    LocalTime max = LocalTime.of(request.getFilterTrips().getMaxHour(),0);
                    System.out.println("Se cauta excursiile in locul: "+s+" incepand cu ora: "+min+"si max ora: "+max);
                    Iterable<Trip> ftrips = server.findTripsByNameAndHours(s, min,max);
                    for(Trip t :ftrips){
                        System.out.println("Excursiile filtrate sunt: "+t);
                    }
                    return ProtoUtils.createFilteredTripsResponse(ftrips);
                } catch (ServiceException e) {
                    return ProtoUtils.createErrorResponse(e.getMessage());
                }
            }
            case ADD_RESERVATION -> {
                System.out.println("New reservation request..");
                Reservation r = ProtoUtils.getReservation(request);
                try{
                    server.addReservation(r);
                    return ProtoUtils.createOkResponse();
                } catch (ServiceException | RepoException | ValidationException e) {
                    return ProtoUtils.createErrorResponse(e.getMessage());
                }
            }

        }
        return response;
    }




    private void sendResponse(AgencyProtobufs.AgencyResponse response) throws IOException{
        System.out.println("sending response "+response);
        response.writeDelimitedTo(output);
        //output.writeObject(response);
        output.flush();
    }
}

File: AgentieTurism/protobuf_3/AgencyProtoWorker.cs
using System;
using System.Collections.Generic;
using System.Net.Sockets;
using System.Threading;
using Agency.Protocol;
using Google.Protobuf;
using services;
using Reservation = model.domain.Reservation;

namespace protobuf_3
{
    public class AgencyProtoWorker:IAgencyObserver
    {
        private IAgencyService server;
        private TcpClient connection;

        private NetworkStream stream;
        private volatile bool connected;

        public AgencyProtoWorker(IAgencyService server, TcpClient connection)
        {
            this.server = server;
            this.connection = connection;
            
            try
            {
				
                stream=connection.GetStream();
                connected=true;
            }
            catch (Exception e)
            {
                Console.WriteLine(e.StackTrace);
            }
        }


        public void ReservationAdded(Reservation reservation)
        {
	        Console.WriteLine("Update new reservation...");
	        try
	        {
		        sendResponse(ProtoUtils.createNewReservationResponse(reservation));
	        }
	        catch (Exception e)
	        {
		        Console.WriteLine(e.StackTrace);
	        }
        }

        public virtual void run()
        {
            while(connected)
            {
                try
                {
					
                    AgencyRequest request = AgencyRequest.Parser.ParseDelimitedFrom(stream);
                    AgencyResponse response =handleRequest(request);
                    if (response!=null)
                    {
                        sendResponse(response);
                    }
                }
                catch (Exception e)
                {
                    Console.WriteLine(e.StackTrace);
                }
				
                try
                {
                    Thread.Sleep(1000);
                }
                catch (Exception e)
                {
                    Console.WriteLine(e.StackTrace);
                }
            }
            try
            {
                stream.Close();
                connection.Close();
            }
            catch (Exception e)
            {
                Console.WriteLine("Error "+e);
            }
        }
        private AgencyResponse handleRequest(AgencyRequest request)
		{
			AgencyResponse response =null;
			AgencyRequest.Types.Type reqType=request.Type;
            switch(reqType){
                case AgencyRequest.Types.Type.Login:
			{
				Console.WriteLine("Login request ...");
				model.domain.TravelAgent user =ProtoUtils.getUser(request);
				try
                {
                    lock (server)
                    {
                        server.Login(user, this);
                    }
					return ProtoUtils.createOkResponse();
				}
				catch (ServiceException e)
				{
					connected=false;
					return ProtoUtils.createErrorResponse(e.Message);
				}
			}
                case AgencyRequest.Types.Type.Logout:
			{
				Console.WriteLine("Logout request");
                model.domain.TravelAgent user = ProtoUtils.getUser(request);
				try
				{
                    lock (server)
                    {

                        server.Logout(user, this);
                    }
					connected=false;
					return ProtoUtils.createOkResponse();

				}
				catch (ServiceException e)
				{
				   return ProtoUtils.createErrorResponse(e.Message);
				}
			}
                case AgencyRequest.Types.Type.AddUser:
			{
				Console.WriteLine("Add user request ...");
				model.domain.TravelAgent user =ProtoUtils.getUser(request);
				try
				{
					lock (server)
					{
						server.AddUser(user);
					}
					return ProtoUtils.createOkResponse();
				}
				catch (ServiceException e)
				{
					return ProtoUtils.createErrorResponse(e.Message);
				}
			}

                case AgencyRequest.Types.Type.GetAllTrips:
			{
				Console.WriteLine("Get all trips request Request ...");
				try
				{
                    IEnumerable<model.domain.Trip> trips;
                    lock (server)
                    {

                         trips = server.GetAllTrips();
                    }
					return ProtoUtils.createGetAllTripsResponse(trips);
				}
				catch (ServiceException e)
				{
					return ProtoUtils.createErrorResponse(e.Message);
				}
			}
                case AgencyRequest.Types.Type.TripsByNameHour:
                {
	                string place = request.FilterTrips.Place;
	                int min = request.FilterTrips.MinHour;
	                int max = request.FilterTrips.MaxHour;
	                Console.WriteLine("Get filtered trips request Request ...");
	                
	                try
	                {
		                IEnumerable<model.domain.Trip> trips;
		                lock (server)
		                {

			                trips = server.FindTripsByNameAndHours(place,min,max);
		                }
		                return ProtoUtils.createFilteredTripsResponse(trips);
	                }
	                catch (ServiceException e)
	                {
		                return ProtoUtils.createErrorResponse(e.Message);
	                }
                }  
                case AgencyRequest.Types.Type.AddReservation:
                {
	                Console.WriteLine("Add user request ...");
	                model.domain.Reservation reservation =ProtoUtils.getReservation(request);
	                try
	                {
		                lock (server)
		                {
			                server.AddReservation(reservation);
		                }
		                return ProtoUtils.createOkResponse();
	                }
	                catch (ServiceException e)
	                {
		                return ProtoUtils.createErrorResponse(e.Message);
	                }
                }
            }
			return response;
		}
        private void sendResponse(AgencyResponse response)
        {
            Console.WriteLine("sending response "+response);
            lock (stream)
            {
                response.WriteDelimitedTo(stream);
                stream.Flush();
            }

        }
    }
    
}